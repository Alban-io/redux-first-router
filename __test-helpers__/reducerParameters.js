import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
