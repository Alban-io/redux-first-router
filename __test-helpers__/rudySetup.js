import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
