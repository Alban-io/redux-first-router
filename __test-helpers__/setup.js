import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
