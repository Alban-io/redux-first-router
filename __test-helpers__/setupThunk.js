import { createStore, applyMiddleware, compose, combineReducers } from 'redux'
import connectRoutes from '../src/connectRoutes'

export default (path = '/', thunkArg, opts, dispatchFirstRoute = true) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk: thunkArg },
    THIRD: { path: '/third/:param' }
  }

  const options = { extra: { arg: 'extra-arg' }, initialEntries: path, ...opts }

  const { middleware, enhancer, reducer, history, firstRoute } = connectRoutes(
    routesMap,
    options
  )

  const rootReducer = combineReducers({
    location: reducer
  })

  const middlewares = applyMiddleware(middleware)
  const enhancers = compose(enhancer, middlewares)
  const store = createStore(rootReducer, enhancers)

  if (dispatchFirstRoute) store.dispatch(firstRoute())

  return { store, firstRoute, history }
}
