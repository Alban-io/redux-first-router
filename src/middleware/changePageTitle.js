import changePageTitle from '../pure-utils/changePageTitle'
import isServer from '../pure-utils/isServer'

export default async (req, next) => {
  const title = req.getTitle()

  if (!isServer() && typeof title === 'string') {
    window.document.title = title
  }

  return next()
}
