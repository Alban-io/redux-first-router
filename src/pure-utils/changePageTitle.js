// @flow
import isServer from './isServer'

import type { Document } from '../flow-types'

export default (doc: Document, title: ?string): ?string => {
  if (typeof title === 'string' && doc.title !== title) {
    return (doc.title = title)
  }

  return null
}

export const getDocument = (): Document => (!isServer() ? document : {})
