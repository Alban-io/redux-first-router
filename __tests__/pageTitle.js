import { setupAll } from '../__test-helpers__/setup'
import changePageTitle from '../src/pure-utils/changePageTitle'

it('title and location options as selector functions', () => {
  const { store } = setupAll('/first', {
    title: state => state.title,
    location: state => state.location
  })

  const action = store.dispatch({ type: 'FIRST' }) /*? $.meta */

  store.getState() /*? $.location */

  expect(action).toMatchSnapshot()
})

it('unit: when title changes set it to document.title', () => {
  const document = {}
  const title = 'foo'

  const ret = changePageTitle(document, title)

  expect(document).toEqual({ title: 'foo' })
  expect(ret).toEqual('foo')
})

it('unit: hen title changes do not set document.title', () => {
  const document = { title: 'foo' }
  const title = 'foo'

  const ret = changePageTitle(document, title)

  expect(document).toEqual({ title: 'foo' })
  expect(ret).toEqual(null) // no return value when title does not change
})
