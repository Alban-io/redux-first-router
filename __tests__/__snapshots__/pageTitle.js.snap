// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`title and location options as selector functions 1`] = `
Object {
  "kind": "push",
  "meta": Object {
    "location": Object {
      "current": Object {
        "kind": "push",
        "pathname": "/first",
        "payload": Object {},
        "type": "FIRST",
      },
      "history": undefined,
      "kind": "push",
      "prev": Object {
        "kind": "load",
        "pathname": "/first",
        "payload": Object {},
        "type": "FIRST",
      },
    },
  },
  "payload": Object {},
  "type": "FIRST",
}
`;
